File: candy.py
class Solution(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        if not ratings:
            return 0

        m = len(ratings)
        i = 0
        buf = [1] * m
        while i < m:
            j = i
            while i + 1 < m and ratings[i + 1] <= ratings[i]:
                i += 1

            buf[i] = 1
            for t in range(i-1, j-1, -1):
                if ratings[t] > ratings[t+1]:
                    buf[t] = buf[t+1] + 1
                else:
                    buf[t] = buf[t+1]
            print buf

            j = i
            while i + 1 < m and ratings[i + 1] >= ratings[i]:
                i += 1

            for t in range(j+1, i+1):
                if ratings[t] > ratings[t-1]:
                    buf[t] = buf[t-1] + 1
                else:
                    buf[t] = buf[t-1]
        return sum(buf)

        # buf = [1]
        # for i in range(1, len(ratings)):
        #     if ratings[i] == ratings[i-1]:
        #         buf.append(buf[-1])
        #     elif ratings[i] < ratings[i-1]:
        #         last = buf[-1]
        #         if last > 1:
        #             buf.append(last - 1)
        #         else:
        #             buf[-1] = buf[-1] + 1
        #             buf.append(buf[-1] - 1)
        #             for j in range(i-2, -1, -1):
        #                 if (ratings[j] >= ratings[j+1] and buf[j] < buf[j+1]) or (
        #                     ratings[j] > ratings[j+1] and buf[j] == buf[j+1]):
        #                     buf[j] = buf[j] + 1

        #     else:
        #         buf.append(buf[-1] + 1)
        # return sum(buf)

if __name__ == '__main__':
    sol = Solution()

    ratings = [3, 2, 10, 4, 5]
    print sol.candy(ratings)

File: candy.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int candy(vector<int>& ratings) {
		int n = ratings.size();
		std::vector<int> v(n, 0);
		v[0] = 1;
		for(int i = 1; i < n; ++i)
		{
			if (ratings[i] > ratings[i-1])
				v[i] = v[i-1] + 1;
			else if (ratings[i] < ratings[i-1])
				v[i] = 1;
			else
				v[i] = 1;
			if (v[i] <= 0)
				v[i] = 1;
		}
		for(int i = n-2; i >= 0; --i)
		{
			if (ratings[i] > ratings[i+1]) {
				if (v[i] <= v[i+1])
					v[i] = v[i+1] + 1;
			}
			else if (ratings[i] < ratings[i+1]) {
				if (v[i] >= v[i+1])
					v[i] = v[i+1] - 1;
			}
			else {
				if (v[i] < 1)
					v[i] = 1;
			}
		}
		int ret = 0;
		for(size_t j = 0; j < v.size(); ++j)
		{
			cout << v[j] << ",";
			ret += v[j];
		}
		cout << endl;
        return ret;
    }
};

int main (int argc, char const *argv[])
{
	Solution s;
//	int ratings[] = {1,2,3,2,1,0,-1,0,1,2,3,4,5,4,3};
	int ratings[] = {1,2,2};
	std::vector<int> v(ratings, ratings+sizeof(ratings)/sizeof(int));
	cout << s.candy(v);
	return 0;
}